### ISSUE TITLE
Unexpected end of file from server exception when several test methods within one class

### ISSUE THREAD

**Issue by tongguop** [tomakehurst/wiremock#132]
This doesn't seem to be a problem on my own computer. However when I run the maven build on Jenkins, it always fails with the following message:
Caused by: java.net.SocketException: SocketException invoking http://localhost:8098/some/useless/path: Unexpected end of file from server.

I searched Wiremock's config, it doesn't seem that I can change the server timeout settings.

**Comment by kevinhooke** [comment]
Regarding the 202 response breaking subsequent calls - if you mock a 204 (no content) response and mistakenly attempt to return a body as well, the next mocked call fails as it appears the previous body is being returned in the following mocked response and it's not correctly formed. I found this thread/issue when I was trying to work out what was going wrong in my test, but in my case, removing the body for my 204 mocked response solved my issue.

**Comment by trzaszcs** [comment]
Hi,

we're encountering the same issue. It occurs from time to time.  So one of the tests fails. We're using RestTemplate (PUT method)

Wiremock is configured like this:

```
@Rule
 WireMockRule wireMockRule = new WireMockRule(12346)
```


and the stub

```
stubFor(put(urlEqualTo("/offers/${offerId}/buy-now-price"))
                .willReturn(aResponse().withStatus(200)
                .withHeader("Content-Type", "application/json")
                .withBody("""{"value":"123456"}""")))
```

The exception 

> ResourceAccessException: I/O error on PUT request for "http://localhost:12346/offers/offerId/buy-now-price": Unexpected end of file from server; nested exception is java.net.SocketException: Unexpected end of file from server

The stub is inside given part (spock)

JDK 1.8
wiremock: 2.1.11

**Comment by tomakehurst** [comment]
Can you post the whole test source code?

**Comment by tomakehurst** [comment]
I'm guessing that example is Groovy rather than Java? If so, by putting 'offerId' in the URL are you making it not match?

**Comment by trzaszcs** [comment]
Sure,

I've simplified test

`class ChangePriceEndpointIntegrationSpec extends IntegrationSpec {

    @Autowired
    OfferServiceClient offerServiceClient

    def "should handle event"() {
        given:
        def itemId = "offerId"
        OfferManagerStub.stub(itemId)
        when:
        offerServiceClient.updateOfferPrice(new OfferUpdatePriceRequest(itemId, 22, 1, "PLN", "TOKEN"))
        then:
        true
    }

}`

**Comment by trzaszcs** [comment]
The logs from the integrationTest

> 13:02:01.391 [Test worker] INFO  o.s.t.c.s.AbstractContextLoader - Could not detect default resource locations for test class [xxx.sale.merchandising.offerproductpriceservice.IntegrationSpec]: no resource found for suffixes {-context.xml, Context.groovy}.
13:02:01.406 [Test worker] INFO  o.s.b.t.c.SpringBootTestContextBootstrapper - Loaded default TestExecutionListener class names from location [META-INF/spring.factories]: [org.springframework.boot.test.mock.mockito.MockitoTestExecutionListener, org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener, org.springframework.boot.test.autoconfigure.AutoConfigureReportTestExecutionListener, org.springframework.boot.test.autoconfigure.restdocs.RestDocsTestExecutionListener, org.springframework.boot.test.autoconfigure.web.client.MockRestServiceServerResetTestExecutionListener, org.springframework.test.context.web.ServletTestExecutionListener, org.springframework.test.context.support.DirtiesContextBeforeModesTestExecutionListener, org.springframework.test.context.support.DependencyInjectionTestExecutionListener, org.springframework.test.context.support.DirtiesContextTestExecutionListener, org.springframework.test.context.transaction.TransactionalTestExecutionListener, org.springframework.test.context.jdbc.SqlScriptsTestExecutionListener]
13:02:01.407 [Test worker] INFO  o.s.b.t.c.SpringBootTestContextBootstrapper - Using TestExecutionListeners: [org.springframework.test.context.web.ServletTestExecutionListener@7b21acf3, org.springframework.test.context.support.DirtiesContextBeforeModesTestExecutionListener@c84c8bf, org.springframework.boot.test.autoconfigure.AutoConfigureReportTestExecutionListener@5a9749d, org.springframework.test.context.support.DependencyInjectionTestExecutionListener@4a4e192c, org.springframework.test.context.support.DirtiesContextTestExecutionListener@3cfe72a2, org.springframework.test.context.transaction.TransactionalTestExecutionListener@169584b2, org.springframework.test.context.jdbc.SqlScriptsTestExecutionListener@58560761, org.springframework.boot.test.mock.mockito.ResetMocksTestExecutionListener@626bf23e, org.springframework.boot.test.mock.mockito.MockitoTestExecutionListener@2dcfe7b3, org.springframework.boot.test.autoconfigure.restdocs.RestDocsTestExecutionListener@174546ac, org.springframework.boot.test.autoconfigure.web.client.MockRestServiceServerResetTestExecutionListener@7fc0262a]
13:02:01.437 [Test worker] INFO  org.eclipse.jetty.server.Server - jetty-9.2.13.v20150730
13:02:01.439 [Test worker] INFO  o.e.j.server.handler.ContextHandler - Started o.e.j.s.ServletContextHandler@21246617{/__admin,null,AVAILABLE}
13:02:01.441 [Test worker] INFO  o.e.j.server.handler.ContextHandler - Started o.e.j.s.ServletContextHandler@60eb4bda{/,null,AVAILABLE}
13:02:01.441 [Test worker] INFO  o.e.jetty.server.ServerConnector - Started ServerConnector@627d9d0e{HTTP/1.1}{0.0.0.0:12347}
13:02:01.441 [Test worker] INFO  org.eclipse.jetty.server.Server - Started @17330ms
13:02:01.473 [qtp1399165453-88] INFO  /__admin - RequestHandlerClass from context returned com.github.tomakehurst.wiremock.http.AdminRequestHandler. Normalized mapped under returned 'null'
13:02:01.528 [Test worker] INFO  p.a.s.m.o.i.o.OfferServiceRetriableClient - Updating offer price:
13:02:01.531 [Test worker] INFO  p.a.s.t.o.OfferManagerClient - Update buyNow price request: ChangePriceDto{price=PriceDto{amount=22, currency=PLN}, userId='1'}, offerId: offerId
13:02:01.580 [Test worker] INFO  o.e.jetty.server.ServerConnector - Stopped ServerConnector@627d9d0e{HTTP/1.1}{0.0.0.0:12347}
13:02:01.581 [Test worker] INFO  o.e.j.server.handler.ContextHandler - Stopped o.e.j.s.ServletContextHandler@60eb4bda{/,null,UNAVAILABLE}
13:02:01.583 [Test worker] INFO  o.e.j.server.handler.ContextHandler - Stopped o.e.j.s.ServletContextHandler@21246617{/__admin,null,UNAVAILABLE}
13:02:01.584 [Test worker] WARN  o.e.j.util.thread.QueuedThreadPool - qtp1399165453{STOPPING,8<=8<=10,i=2,q=4} Couldn't stop Thread[qtp1399165453-84,5,main]
13:02:01.584 [Test worker] WARN  o.e.j.util.thread.QueuedThreadPool - qtp1399165453{STOPPING,8<=8<=10,i=2,q=4} Couldn't stop Thread[qtp1399165453-89,5,main]

**Comment by tomakehurst** [comment]
Sorry, should've been more specific. Can you post the code that starts the WireMock server and the actual HTTP client code that sends the PUT request?

**Comment by trzaszcs** [comment]
Please keep in mind that if you run single test from IDE it's 'GREEN'

If there are more tests that interacts with Wiremock the problem occurs. The application is Spring Boot app

**Comment by tomakehurst** [comment]
I'm going to struggle to replicate this with the info you've given me so far. Happy to investigate but only if I can get to a failing test.

**Comment by trzaszcs** [comment]
Hello,

I've simplified test:

`class ChangePriceEndpointIntegrationSpec extends Specification {

    @Rule
    WireMockRule wireMockRule = new WireMockRule(12346)

    RestTemplate restTemplate = new RestTemplate()

    def "should handle event"() {
        given:
        def itemId = "offerId"
        stubFor(put(urlEqualTo("/offers/${itemId}/buy-now-price"))
                .willReturn(aResponse().withStatus(200)
                .withHeader("Content-Type", "application/json")
                .withBody("""{"value":"123456"}""")))
        RequestEntity<ChangePriceDto> request = new RequestEntity<>(new ChangePriceDto(), new LinkedMultiValueMap<>(), HttpMethod.PUT, new URI("http://localhost:12346/offers/${itemId}/buy-now-price"))
        when:
        ResponseEntity<JobIdDto> response = restTemplate.exchange(request, JobIdDto.class);
        then:
        println response
        true
    }

}`

There are other integration tests that passes and uses WireMockRule on the same port
As I mentioned. Running test from IDE it's fine. If I rune all tests from gradle - it passes from time to time

**Comment by tomakehurst** [comment]
Which WireMock version are you running?

**Comment by trzaszcs** [comment]
I'm investigating issue on my side

I've changed the code a little bit. 
I've added groovy [Http-builder](https://github.com/jgritman/httpbuilder) client.
Groovy client is working fine, RestTemplate still causes the problem ? 

Modified code:


`
    class ChangePriceEndpointIntegrationSpec extends Specification {

    @Rule
    WireMockRule wireMockRule = new WireMockRule(12346)

    def "should handle event"() {
        given:
        def itemId = "offerId"
        // here is stub
        stubFor(put(urlEqualTo("/offers/${itemId}/buy-now-price"))
                .willReturn(aResponse().withStatus(200)
                .withHeader("Content-Type", "application/json")
                .withBody("""{"value":"123456"}""")))

        // init clients
        RestTemplate restTemplate = new RestTemplate()
        HTTPBuilder http = new HTTPBuilder("http://localhost:12346/offers/${itemId}/buy-now-price");

        MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
        headers.put(HttpHeaders.ACCEPT, Lists.newArrayList("application/json"));
        RequestEntity<ChangePriceDto> request = new RequestEntity<>(new ChangePriceDto(), headers, HttpMethod.PUT, new URI("http://localhost:12346/offers/${itemId}/buy-now-price"))

        when:
        http.request(PUT, TEXT) { req ->
            body = [ticket: [subject: 'xyz']]

            response.success = { resp, json ->
                println "Success! ${resp.status} ${json}"
            }

            response.failure = { resp ->
                println "Request failed with status ${resp.status} ${resp.responseData}"
            }
        }
        ResponseEntity<JobIdDto> response = restTemplate.exchange(request, JobIdDto.class);
        then:
        println response
    }

}
`

**Comment by trzaszcs** [comment]
wiremock: 2.1.11

**Comment by tomakehurst** [comment]
OK, rules out the problem I thought it might have been. I guess the signs are that it's a RESTTemplate problem in any case. Happy to take a look again if you find anything that points to WM being at issue.

**Comment by trzaszcs** [comment]
It looks like race condition.

I have 2 tests.

1 test invokes Rest endpoint that calls WM

2 test:

`class ChangePriceEndpointIntegrationSpec extends Specification {

    @Rule
    WireMockRule wireMockRule = new WireMockRule(12346)

    def "should handle event"() {
        given:
        // here is stub
        stubFor(put(urlEqualTo("/some"))
                .willReturn(aResponse().withStatus(200)
                .withHeader("Content-Type", "application/json")
                .withBody("""{"value":"123456"}""")))

        // init client
        RestTemplate restTemplate = new RestTemplate()
        MultiValueMap<String, String> headers = new LinkedMultiValueMap<>();
        headers.put(HttpHeaders.ACCEPT, Lists.newArrayList("application/json"));
        RequestEntity<ChangePriceDto> request = new RequestEntity<>(new ChangePriceDto(), headers, HttpMethod.PUT, new URI("http://localhost:12346/some"))

//        sleep(5000)

        when:
        2.times {
            try {
                println restTemplate.exchange(request, JobIdDto.class);
            } catch (Exception ex) {
                println "failure ${ex}"
            }
        }
        then:
        "OK"
    }

}`


First call fails, second is fine. If you add sleep(5000) both calls are working.

Most intereting part are logs. I don't know WM so good so maybe You can help me with that

> 10:00:59.968 [main] INFO  o.e.jetty.server.ServerConnector - Started ServerConnector@48697181{HTTP/1.1}{0.0.0.0:12346}
> 10:00:59.970 [main] INFO  org.eclipse.jetty.server.Server - Started @13654ms
> 10:00:59.978 [qtp237594516-40] INFO  /__admin - RequestHandlerClass from context returned com.github.tomakehurst.wiremock.http.AdminRequestHandler. Normalized mapped under returned 'null'
> failure org.springframework.web.client.ResourceAccessException: I/O error on PUT request for "http://localhost:12346/some": Unexpected end of file from server; nested exception is java.net.SocketException: Unexpected end of file from server
> 10:01:00.057 [qtp237594516-41] INFO  / - RequestHandlerClass from context returned com.github.tomakehurst.wiremock.http.StubRequestHandler. Normalized mapped under returned 'null'
> <200 OK,JobIdDto{value='123456'},{Content-Type=[application/json], Transfer-Encoding=[chunked], Server=[Jetty(9.2.13.v20150730)]}>
> 10:01:00.093 [main] INFO  o.e.jetty.server.ServerConnector - Stopped ServerConnector@48697181{HTTP/1.1}{0.0.0.0:12346}


Interesting is this entry :

> 10:01:00.057 [qtp237594516-41] INFO  / - RequestHandlerClass from context returned com.github.tomakehurst.wiremock.http.StubRequestHandler. Normalized mapped under returned 'null'

It occurs just after failure ...

**Comment by trzaszcs** [comment]
Hello,
if these logs are not helpful - please let me know. Issue is quite annoying

Thanks

**Comment by tomakehurst** [comment]
It's not obvious what's happening from the logs. I think I'm going to have to set up my own concurrent test and hope that I can replicate it. Agree it looks like a race condition, but the question is where it exists.

**Comment by georgethekkan** [comment]
We had similar problems, Spring rest template invocation always failed in few test cases. 

Looks like this has something to do with Spring restTemplate's underlying HTTP handling. To verify that this we added a Apache httpClient rest invocation before spring and httpClient invocations were always success while restTemplate is failing. 

We have solved this by using the Apache httpClient as the RestTemplate.setRequestFactory()

**Comment by tomakehurst** [comment]
That's very useful to know. Thanks for sharing!

**Comment by shivam-maharshi** [comment]
This issue is still occurring with 2.2 version, using ClassRules and no other framework even for a single HTTP Get request too. For now I am resorting to writing an embedded tomcat server for testing. Can we expect a solution soon?

**Comment by anicos** [comment]
For me works good after changing ClientHttpRequestFactory like this `RestTemplate restTemplate = new RestTemplate(new HttpComponentsClientHttpRequestFactory());`

**Comment by trzaszcs** [comment]
Hello,
solution proposed by @anicos didn't help me. It looks like the problem is harder to solve.

**Comment by trzaszcs** [comment]
Ok I've got something. Maybe the problem still occurs. But I prepared separate RestTemplate for RestTemplate.

Related problem and solution is reported [here](http://stackoverflow.com/a/10680629)

Here is the code 

`import org.apache.http.client.HttpRequestRetryHandler;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.protocol.HttpContext;
import org.springframework.context.annotation.Bean;
import org.springframework.http.client.HttpComponentsClientHttpRequestFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import java.io.IOException;

@Component
public class TestRestConfig {

    @Bean
    RestTemplate seachClient() {
        return new RestTemplate(new HttpComponentsClientHttpRequestFactory(HttpClientBuilder.create().setRetryHandler(new HttpRequestRetryHandler() {
            @Override
            public boolean retryRequest(IOException exception, int executionCount, HttpContext context) {
                return true;
            }
        }).build()));
    }

}`

**Comment by tomakehurst** [comment]
OK, I'm getting pretty confused by this thread now. I created this test - executing PUTs using a default RestTemplate across 200 threads:
https://gist.github.com/tomakehurst/dd09638e631acf76e723c18fa52fb417

I can't get this to fail.

If anybody on here still believes this is a WireMock bug, please can they provide a full test project on GH I can check out and run, that reliably fails. Otherwise I'm going to assume this is a client bug and close it off.

**Comment by anicos** [comment]
@trzaszcs you are right my solutions doesn't work. My mistake. sorry:(

**Comment by anicos** [comment]
Finally I have solution by workaround. In my project I use random port for WireMock
`@Rule
     WireMockRule wireMockRule = new WireMockRule(Options.DYNAMIC_PORT)`

And it is working prefect. It's even better solution because now I can run my tests in parallel:)

**Comment by tlefevre** [comment]
Having the same issue in our project. I switched from using the Unirest client to pure HttpUrlConnection and started having issues with interactions with WireMock failing. 

The first test will run smoothly, but the second one will always fail. If I reverse the execution order of the test, the problem stays the same. So there is no inherent problem in the test.

I can fix the issue quite simply via a Thread.sleep statement after setting up the new mock that I want to interact with.

**Comment by tomakehurst** [comment]
@tlefevre please could you post a test project that reliably replicates this?

**Comment by sylvainlaurent** [comment]
see #97 https://github.com/tomakehurst/wiremock/issues/97#issuecomment-288983671 for a reproducer

**Comment by vkolotov** [comment]
Same problem here with the most recent version "2.15.0" and even with HttpComponentsClientHttpRequestFactory.

**Comment by addonin** [comment]
> Changing from :
> `@Rule public WireMockRule wireMockRule = new WireMockRule(9091);`
> to
> `@ClassRule public static WireMockRule wireMockRule = new WireMockRule(9091);`
> 
> worked for me.

Helped me too

**Comment by sdoeringNew** [comment]
Having the same problem between test classes.

Already using ``ClassRule`` annotation because of Spring Boots ``SpringRunner``.

Both tests use the same two ``WireMockRule`` with the same ports.

The order of the tests is not relevant.
First ``Test1`` then ``Test2`` and the first test method in ``Test2`` using the rules fails.
First ``Test2`` then ``Test1`` and the first test method in ``Test1`` using the rules fails.

**Comment by tomakehurst** [comment]
@sdoeringNew please can you post a test project with the failing tests?

Also, try these: https://github.com/tomakehurst/wiremock/issues/97#issuecomment-496681952

**Comment by sdoeringNew** [comment]
@tomakehurst If you don't mind that it is a Maven project, then yes.
I've just created a simple test application those tests are failing with the error: https://github.com/sdoeringNew/oh_noez_wiremock

I'll also look at the comment.

**Comment by sdoeringNew** [comment]
> Ensure WireMock is starting and stopping in unison with the app under test. In practice this usually means using @ClassRule in JUnit rather than @Rule with the WireMockRule.

Already using ``@ClassRule`` and the error occurs.

> Disable chunked transfer encoding (see http://wiremock.org/docs/configuration/#transfer-encoding)

No change.

> Switch away from Java's HttpURLConnection as an HTTP client, as it has buggy connection pool handling in some circumstances. Note: many HTTP clients, including Spring's RestTemplate and the Jersey client will use this under the hood by default.

Already done that and the error would be gone.

> If you're using the Netty HttpClient or something that wraps it such as Spring's reactive WebClient, ensure you have the latest version of the Netty dependency. See #914 (comment)

No Netty.

> Turn off connection reuse in your HTTP client.

With the system property ``http.keepAlive=false`` the error is gone but a warning including stacktrace occurs:

    java.nio.channels.ClosedChannelException: null

> If you're using Spring, use org.springframework.cloud:spring-cloud-contract-wiremock (as described in the previous comment). See https://cloud.spring.io/spring-cloud-contract/multi/multi__spring_cloud_contract_wiremock.html

Same error.

**Comment by tomakehurst** [comment]
@sdoeringNew  thanks for posting the test project.

When I run all tests I get an immediate connection reset error in test2:
```
feign.RetryableException: Connection reset executing POST http://localhost:8888/test
	at com.example.wiremock.Test1Test.test1(Test1Test.java:39)
Caused by: java.net.SocketException: Connection reset
	at com.example.wiremock.Test1Test.test1(Test1Test.java:39)
```

Is this also what you see?

**Comment by tomakehurst** [comment]
Ah OK. Feign's default underlying client is `HTTPUrlConnection`, which would explain why this is happening.

`HTTPUrlConnection` will pool connections to `localhot:8888` between tests unless you set `http.keepAlive=false`. So you will end up with a reference to a dead connection in the second test, since WireMock has been shut down and restarted in the meantime.

More advanced HTTP clients will detect this and create a fresh connection but `HTTPUrlConnection` won't. I've looked at this variation of the problem in the past, but I can't see any way this can be fixed at the WireMock end.

**Comment by jgroeneveld** [comment]
We configured wiremock to return
```
      "Connection": "close"
```
header. This also worked for us. Obviously not a great solution..

**Comment by tomakehurst** [comment]
I'll mention this again - if you can possibly avoid using `HTTPURLConnection` (directly or as an underlying impl) then I'd say you should do that. It's not very good, and nearly always at the root of these kinds of problems.

**Comment by jgroeneveld** [comment]
> I'll mention this again - if you can possibly avoid using `HTTPURLConnection` (directly or as an underlying impl) then I'd say you should do that. It's not very good, and nearly always at the root of these kinds of problems.

Agreed, sadly not a choice when a third party library uses it.

**Comment by sdoeringNew** [comment]
> I'll mention this again - if you can possibly avoid using `HTTPURLConnection` (directly or as an underlying impl) then I'd say you should do that. It's not very good, and nearly always at the root of these kinds of problems.

Especially in JUnit tests or generally?

**Comment by tomakehurst** [comment]
@sdoeringNew generally, I'd say. It's buggy and has very limited control over connection lifecycle.

**Comment by tomakehurst** [comment]
Closing this as it's old, and not really a WireMock issue.
